compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.h

```diff
@@ -38,6 +41,76 @@ LogicalResult foldUnitDims(const SmallVector<OpFoldResult> &offsets,
                            SmallVector<OpFoldResult> &newStrides,
                            SmallVector<OpFoldResult> &newSizes);
 
+/// Utility to discard all non-zero offsets that have dimension equal to 1 on
+/// the same index of the provided shape. This helps with updating DMA
+/// operations for a shape change.
+template <CopyOpOperateOn OperateOn>
+AMDAIE::DoublyStridedOpInterface discardAllNonZeroOffsets(
+    RewriterBase &rewriter, AMDAIE::DoublyStridedOpInterface op,
+    ArrayRef<int64_t> shape = {}) {
+  SmallVector<OpFoldResult> newSourceOffsets;
+  SmallVector<OpFoldResult> newSourceSizes;
+  SmallVector<OpFoldResult> newSourceStrides;
+  SmallVector<OpFoldResult> newTargetOffsets;
+  SmallVector<OpFoldResult> newTargetSizes;
+  SmallVector<OpFoldResult> newTargetStrides;
+  if constexpr (OperateOn == CopyOpOperateOn::Source) {
+    SmallVector<OpFoldResult> offsets = op.getSourceMixedOffsets();
+    SmallVector<OpFoldResult> sizes = op.getSourceMixedSizes();
+    SmallVector<OpFoldResult> strides = op.getSourceMixedStrides();
+    // Set shape to a vector of ones as a default.
+    if (shape.empty()) {
+      SmallVector<int64_t> ones(offsets.size(), 1);
+      shape = ones;
+    }
+    if (shape.size() != offsets.size()) return op;
+    // Fill source offsets/sizes/strides.
+    for (auto &&[offset, size, stride, dim] :
+         llvm::zip(offsets, sizes, strides, shape)) {
+      std::optional<int64_t> constantOffset = getConstantIntValue(offset);
+      if (dim == 1 &&
+          (!constantOffset || (constantOffset && constantOffset.value() != 0)))
+        continue;
```

Comment:
Can you break this into several if statements? We normally don't want to mix `&&` and `||` conditions.

Reply:
Done